src/ddmd/expressionsem.d

```diff
@@ -68,6 +68,271 @@ import ddmd.visitor;
+        }
+        {
+            FuncDeclaration fd = resolveFuncCall(loc, sc, s, tiargs, tthis, null, 1);
+            if (fd && fd.type)
+            {
+                if (fd.errors)
+                    return new ErrorExp();
+                assert(fd.type.ty == Tfunction);
+                TypeFunction tf = cast(TypeFunction)fd.type;
+                if (!e2 || tf.isref)
+                {
+                    Expression e = new CallExp(loc, e1);
+                    if (e2)
+                        e = new AssignExp(loc, e, e2);
+                    return e.expressionSemantic(sc);
+                }
+            }
+        }
+        if (FuncDeclaration fd = s.isFuncDeclaration())
+        {
+            // Keep better diagnostic message for invalid property usage of functions
+            assert(fd.type.ty == Tfunction);
+            TypeFunction tf = cast(TypeFunction)fd.type;
+            Expression e = new CallExp(loc, e1, e2);
+            return e.expressionSemantic(sc);
+        }
+        if (e2)
+            goto Leprop;
+    }
+    if (e1.op == TOKvar)
+    {
+        VarExp ve = cast(VarExp)e1;
+        VarDeclaration v = ve.var.isVarDeclaration();
+        if (v && ve.checkPurity(sc, v))
+            return new ErrorExp();
+    }
+    if (e2)
+        return null;
+
+    if (e1.type && e1.op != TOKtype) // function type is not a property
+    {
+        /* Look for e1 being a lazy parameter; rewrite as delegate call
+         */
+        if (e1.op == TOKvar)
+        {
+            VarExp ve = cast(VarExp)e1;
+            if (ve.var.storage_class & STClazy)
+            {
+                Expression e = new CallExp(loc, e1);
+                return e.expressionSemantic(sc);
+            }
+        }
+        else if (e1.op == TOKdotvar)
+        {
+            // Check for reading overlapped pointer field in @safe code.
+            if (checkUnsafeAccess(sc, e1, true, true))
+                return new ErrorExp();
+        }
+        else if (e1.op == TOKdot)
+        {
+            e1.error("expression has no value");
+            return new ErrorExp();
+        }
+        else if (e1.op == TOKcall)
+        {
+            CallExp ce = cast(CallExp)e1;
+            // Check for reading overlapped pointer field in @safe code.
+            if (checkUnsafeAccess(sc, ce.e1, true, true))
+                return new ErrorExp();
+        }
+    }
+
+    if (!e1.type)
+    {
+        error(loc, "cannot resolve type for %s", e1.toChars());
+        e1 = new ErrorExp();
+    }
+    return e1;
+
+Leprop:
+    error(loc, "not a property %s", e1.toChars());
+    return new ErrorExp();
+
+Leproplvalue:
+    error(loc, "%s is not an lvalue", e1.toChars());
+    return new ErrorExp();
+}
+
+extern (C++) Expression resolveProperties(Scope* sc, Expression e)
+{
+    //printf("resolveProperties(%s)\n", e.toChars());
+    e = resolvePropertiesX(sc, e);
+    if (e.checkRightThis(sc))
+        return new ErrorExp();
+    return e;
+}
+
```

Comment:
Nit: remove extra whitespace.

Reply:
What extra whitespace?

Reply:
Never mind, it looked like there were three empty lines in https://github.com/dlang/dmd/pull/7251/commits/320581bab36aefbf0fbe4719d4a476af535067c5#diff-b90dc1ef3092d86c410bd0a78271e80eR335, but it looks like you fixed this a later commit.